static/js/navigate.js

```diff
@@ -48,8 +50,20 @@ export function to_home() {
 }
 
 export function to_end() {
-    const next_id = message_lists.current.last().id;
     message_viewport.set_last_movement_direction(1);
+    if (
+        narrow_state.narrowed_to_topic() &&
+        !message_lists.current.data.fetch_status.has_found_newest()
+    ) {
+        const narrowed_topic = narrow_state.topic();
+        const narrowed_stream = narrow_state.stream();
+        narrow.activate([
+            {operator: "stream", operand: narrowed_stream},
+            {operator: "topic", operand: narrowed_topic},
+        ]);
```

Comment:
Do we need to return here? I would expect to be able to tell `narrow.activate` that we want to land at the end via the `opts` parameter, and I believe it can sometimes need to od its own callback.

I suspect this implementation won't work as expected when there are unreads in the narrow, since the default `narrow.activate` behavior will go to the first unread, not the end, if there are any unreads.

Reply:
Yeah, `narrow.activate` would go to the first unread and that's why we do not return here. And the below code now has the last message loaded locally and thus it scrolls down to it while marking all as read. Now reading the code, I feel this is somewhat hacky solution and maybe we would need to do something more complex than simply calling `narrow.activate` as also discussed on CZO.

Reply:
The `then_select_id` parameter is supposed to control which message we narrow around; it's used for things like "returning to a specific message ID when the browser reloads".